source/Mlos.Notebooks/BayesianOptimization-Context.ipynb

```diff
@@ -0,0 +1,464 @@
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "You can see that depending on the choice of `y`, the optimum for `x` is in different locations. For y=-1 it is around 0.85, while for y=0 it's around 0.35."
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Our goal here is to find the global minimum of this function for any `y` that is given, assuming that we don't have direct access to the formula (given the formula, we could instead calculate the optimum quite precicely using methods from calculus instead). Usually, the function is too expensive to evaluate in such a manner, in particular in higher-dimensional spaces.\n",
+    "\n",
+    "Now, we use MLOS to construct an OptimizationProblem object that will encapsulate the function and the input space."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 6,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "from mlos.Optimizers.OptimizationProblem import OptimizationProblem, Objective\n",
+    "from mlos.Optimizers.BayesianOptimizer import BayesianOptimizer\n",
+    "from mlos.Spaces import SimpleHypergrid, ContinuousDimension\n",
+    "\n",
+    "# single continuous input dimension between 0 and 1\n",
+    "input_space = SimpleHypergrid(name=\"input\", dimensions=[ContinuousDimension(name=\"x\", min=0, max=1)])\n",
+    "# define output space, we might not know the exact ranges\n",
+    "output_space = SimpleHypergrid(name=\"objective\",\n",
+    "                               dimensions=[ContinuousDimension(name=\"function_value\", min=-10, max=10)])\n",
+    "# use a context space between -1 and 1\n",
+    "context_space = SimpleHypergrid(name=\"context\", dimensions=[ContinuousDimension(name=\"y\", min=-1, max=1)])\n",
+    "\n",
+    "# define optimization problem with input and output space and objective\n",
+    "optimization_problem = OptimizationProblem(\n",
+    "    parameter_space=input_space,\n",
+    "    objective_space=output_space,\n",
+    "    # we want to minimize the function\n",
+    "    objectives=[Objective(name=\"function_value\", minimize=True)],\n",
+    "    context_space=context_space\n",
+    ")"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "We set up our optimization as before:"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 15,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "from mlos.Optimizers.BayesianOptimizer import BayesianOptimizer, bayesian_optimizer_config_store\n",
+    "from mlos.Optimizers.RegressionModels.HomogeneousRandomForestConfigStore import homogeneous_random_forest_config_store\n",
+    "\n",
+    "from mlos.Spaces import Point\n",
+    "\n",
+    "optimizer_config = bayesian_optimizer_config_store.default\n",
+    "optimizer_config.experiment_designer_config_fraction_random_suggestions = .1\n",
+    "random_forest_config = homogeneous_random_forest_config_store.default\n",
```

Comment:
A bunch of these are the default already.

Reply:
which ones? ``random_forest_config.decision_tree_regression_model_config.n_new_samples_before_refit = 1``? any other ones?